packages/workspace/src/workspace/changed_by_index.ts

```diff
@@ -0,0 +1,191 @@
+    elementIds.push(change.data.after.elemID)
+    authorMap[author] = elementIds
+  }
+}
+
+const updateRemovalChange = async (
+  change: RemovalChange<Element>,
+  authorMap: Record<string, ElemID[]>,
+): Promise<void> => {
+  const author = getChangeAuthor(change)
+  const elementIds = authorMap[author]
+  if (elementIds) {
+    _.remove(elementIds, elemId => elemId.isEqual(change.data.before.elemID))
+    authorMap[author] = elementIds
+  }
+}
+
+const updateModificationChange = async (
+  change: ModificationChange<Element>,
+  authorMap: Record<string, ElemID[]>,
+): Promise<void> => {
+  if (change.data.after.annotations[CORE_ANNOTATIONS.CHANGED_BY]
+    === change.data.before.annotations[CORE_ANNOTATIONS.CHANGED_BY]) {
+    await updateAdditionChange(
+        toChange({ after: change.data.after }) as AdditionChange<Element>,
+        authorMap,
+    )
+  } else {
+    await updateRemovalChange(
+      toChange({ before: change.data.before }) as RemovalChange<Element>,
+      authorMap,
+    )
+    await updateAdditionChange(
+      toChange({ after: change.data.after }) as AdditionChange<Element>,
+      authorMap,
+    )
+  }
+}
+
+const updateChange = async (
+  change: Change<Element>,
+  authorMap: Record<string, ElemID[]>,
+): Promise<void> => {
+  if (isAdditionChange(change)) {
+    await updateAdditionChange(change, authorMap)
+  } else if (isRemovalChange(change)) {
+    await updateRemovalChange(change, authorMap)
+  } else {
+    await updateModificationChange(change, authorMap)
+  }
+}
+
+const getAuthorList = (changes: Change<Element>[]): string[] => {
+  const authorMap: string[] = []
+  const addAuthor = (author: string): void => {
+    if (!authorMap.includes(author)) {
+      authorMap.push(author)
+    }
+  }
+  changes.forEach(change => {
+    if (isModificationChange(change)) {
+      addAuthor(getChangeAuthor(toChange({ before: change.data.before })))
+      addAuthor(getChangeAuthor(toChange({ after: change.data.after })))
+    } else {
+      addAuthor(getChangeAuthor(change))
+    }
+  })
+  return authorMap
+}
+
+const getCompleteAuthorMap = async (
+  changes: Change<Element>[],
+  index: RemoteMap<ElemID[]>,
+): Promise<Record<string, ElemID[]>> => {
+  const authorList = getAuthorList(changes)
+  const indexValues = await index.getMany(authorList)
+  const authorMap: Record<string, ElemID[]> = _.pickBy(
+    _.zipObject(authorList, indexValues),
+    isDefined,
+  )
+  changes.forEach(change => updateChange(change, authorMap))
+  return authorMap
+}
+
+const updateChanges = async (
+  changes: Change<Element>[],
+  index: RemoteMap<ElemID[]>
+): Promise<void> => {
+  const completeAuthorMap = await getCompleteAuthorMap(changes, index)
+  const toBeRemoved = Object.keys(completeAuthorMap).filter(key => isEmpty(completeAuthorMap[key]))
+  await index.setAll(Object.entries(completeAuthorMap).map(([key, value]) => ({ key, value })))
+  await index.deleteAll(toBeRemoved)
```

Comment:
nit: to avoid setting keys that will be deleted right after
```suggestion
  const [toBeRemoved, toBeSet] = _.partition(Object.keys(completeAuthorMap), key => isEmpty(completeAuthorMap[key])
  await index.setAll(Object.entries(toBeSet).map(key => ({ key, value:completeAuthorMap[key] })))
  await index.deleteAll(toBeRemoved)
```

Reply:
nice code